pallets/phala/src/fat.rs

```diff
@@ -0,0 +1,284 @@
+/// Public key registry for workers and contracts.
+pub use self::pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use codec::Encode;
+	use frame_support::{dispatch::DispatchResult, pallet_prelude::*, traits::StorageVersion};
+	use frame_system::pallet_prelude::*;
+	use sp_core::H256;
+	use sp_runtime::traits::Hash;
+	use sp_std::prelude::*;
+	use sp_std::vec;
+
+	use crate::{mq::MessageOriginInfo, registry};
+	// Re-export
+	pub use crate::attestation::{Attestation, IasValidator};
+
+	use phala_types::{
+		contract::messaging::{ContractEvent, ContractOperation},
+		contract::{CodeIndex, ContractClusterId, ContractId, ContractInfo, DeployTarget},
+		messaging::{bind_topic, DecodedMessage, MessageOrigin, WorkerContractReport},
+		ContractPublicKey, WorkerIdentity, WorkerPublicKey,
+	};
+
+	bind_topic!(ContractRegistryEvent, b"^phala/registry/contract");
+	#[derive(Encode, Decode, Clone, Debug)]
+	pub enum ContractRegistryEvent {
+		PubkeyAvailable {
+			contract_id: ContractId,
+			pubkey: ContractPublicKey,
+		},
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+	}
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	/// Mapping from an original code hash to the original code, untouched by instrumentation
+	#[pallet::storage]
+	pub type ContractCode<T: Config> = StorageMap<_, Twox64Concat, CodeHash<T>, Vec<u8>>;
+
+	/// The contract cluster counter.
+	#[pallet::storage]
+	pub type ContractClusterCounter<T> = StorageValue<_, u64, ValueQuery>;
+
+	#[pallet::storage]
+	pub type ContractClusters<T> = StorageMap<_, Twox64Concat, ContractClusterId, Vec<ContractId>>;
+
+	#[pallet::storage]
+	pub type Contracts<T: Config> =
+		StorageMap<_, Twox64Concat, ContractId, ContractInfo<CodeHash<T>, T::AccountId>>;
+
+	#[pallet::storage]
+	pub type ClusterWorkers<T> =
+		StorageMap<_, Twox64Concat, ContractClusterId, Vec<WorkerPublicKey>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		CodeUploaded(CodeHash<T>),
+		PubkeyAvailable(ContractId, ContractPublicKey),
+		Instantiating(ContractId, ContractClusterId, T::AccountId),
+		Instantiated(ContractId, ContractClusterId, H256),
+		InstantiationFailed(ContractId, ContractClusterId, H256),
```

Comment:
Suggest to use struct enum variant. E.g., instead of `CodeUploaded(CodeHash<T>)`, we can write `CodeUploaded { hash: CodeHash<T> }`.

Reply:
Fixed.